### PR TITLE
Windows with gfortran

### LINKED ISSUE #80: SciPy builds for Windows
~~Is it possible to add OpenBLAS builds of scipy for Windows?~~ As noted by @tadeu there is no `scipy` package for Windows in `conda-forge`, so the proper request to make here the addition of (OpenBLAS-enabled) Windows builds of SciPy to `conda-forge`. This would help a lot with the size of executables made from Python programs on Windows.
 
The MKL libraries installed by Anaconda on Windows when installing scipy are very large, so if making an executable (using PyInstaller or similar), even a simple `import scipy; print("Hello, world!")` ends up at around 200 MB.

A similar issue for numpy can be found here: https://github.com/conda-forge/numpy-feedstock/issues/84

### LINKED ISSUE #119: Manual scipy build for win64 py38
I've built scipy manually on an azure vm provisioned with VS2017 + 2019 and Intel IFort.

The only changes to the recipe needed were for compiler activations.

Almost all the tests pass natively aside from the following

```
2 failed, 14587 passed, 1285 skipped, 79 xfailed, 8 xpassed, 40 warnings in 372.07s (0:06:12)
```

```
================================== FAILURES ===================================
__________________ TestDualAnnealing.test_slsqp_ls_minimizer __________________
[gw0] win32 -- Python 3.8.0 %PREFIX%\python.exe

self = <scipy.optimize.tests.test__dual_annealing.TestDualAnnealing testMethod=test_slsqp_ls_minimizer>

    def test_slsqp_ls_minimizer(self):
        minimizer_opts = {
            'method': 'SLSQP',
        }
        ret = dual_annealing(self.func, self.ld_bounds,
                             local_search_options=minimizer_opts)
>       assert_allclose(ret.fun, 0., atol=1e-7)
E       AssertionError:
E       Not equal to tolerance rtol=1e-07, atol=1e-07
E
E       Mismatch: 100%
E       Max absolute difference: 1.67455276e-07
E       Max relative difference: inf
E        x: array(1.674553e-07)
E        y: array(0.)

minimizer_opts = {'method': 'SLSQP'}
ret        =      fun: 1.6745527631201185e-07
 message: ['Maximum number of iteration reached']
    nfev: 4045
    nhev: 0
     nit: 1000
    njev: 8
  status: 0
 success: True
       x: array([2.90527124e-05, 4.65601445e-08])
self       = <scipy.optimize.tests.test__dual_annealing.TestDualAnnealing testMethod=test_slsqp_ls_minimizer>

..\_test_env\Lib\site-packages\scipy\optimize\tests\test__dual_annealing.py:259: AssertionError
____________________ TestQRdelete_F.test_delete_last_p_col ____________________
[gw2] win32 -- Python 3.8.0 %PREFIX%\python.exe

self = <scipy.linalg.tests.test_decomp_update.TestQRdelete_F object at 0x0000019C23FC40A0>

    def test_delete_last_p_col(self):
        a, q, r = self.generate('tall', 'full')
        q1, r1 = qr_delete(q, r, 0, a.shape[1], 'col')
>       assert_unitary(q1)

a          = array([[0.36852294+0.7065659j , 0.9763755 +0.8815727j ,
        0.73179656+0.5390478j , 0.6848747 +0.2577209j ,
      ...0.8069605j ,
        0.2923851 +0.9023696j , 0.08459105+0.7541054j ,
        0.8391204 +0.23152167j]], dtype=complex64)
q          = array([[-0.11519074-0.22085421j,  0.4031076 +0.25545135j,
         0.20641464-0.02725353j,  0.04916074+0.12054591j,
  ....17945503j,  0.10257003-0.11285198j,
        -0.05898733+0.03267026j,  0.5694973 -0.06679238j]],
      dtype=complex64)
q1         = array([[-0.11519074-0.22085421j,  0.4031076 +0.25545135j,
         0.20641464-0.02725353j,  0.04916074+0.12054591j,
  ....17945503j,  0.10257003-0.11285198j,
        -0.05898733+0.03267026j,  0.5694973 -0.06679238j]],
      dtype=complex64)
r          = array([[-3.1992414 +0.j        , -2.1899545 +0.23892963j,
        -1.6467261 +0.47293147j, -2.4033258 +0.41776413j,
  ...        ,
         0.        +0.j        ,  0.        +0.j        ,
         0.        +0.j        ]], dtype=complex64)
r1         = array([], shape=(12, 0), dtype=complex64)
self       = <scipy.linalg.tests.test_decomp_update.TestQRdelete_F object at 0x0000019C23FC40A0>

..\_test_env\Lib\site-packages\scipy\linalg\tests\test_decomp_update.py:327:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = array([[-0.11519074-0.22085421j,  0.4031076 +0.25545135j,
         0.20641464-0.02725353j,  0.04916074+0.12054591j,
  ....17945503j,  0.10257003-0.11285198j,
        -0.05898733+0.03267026j,  0.5694973 -0.06679238j]],
      dtype=complex64)
rtol = 0.0001, atol = 2.384185791015625e-07, assert_sqr = True

    def assert_unitary(a, rtol=None, atol=None, assert_sqr=True):
        if rtol is None:
            rtol = 10.0 ** -(np.finfo(a.dtype).precision-2)
        if atol is None:
            atol = 2*np.finfo(a.dtype).eps

        if assert_sqr:
            assert_(a.shape[0] == a.shape[1], 'unitary matrices must be square')
        aTa = np.dot(a.T.conj(), a)
>       assert_allclose(aTa, np.eye(a.shape[1]), rtol=rtol, atol=atol)
E       AssertionError:
E       Not equal to tolerance rtol=0.0001, atol=2.38419e-07
E
E       Mismatch: 1.39%
E       Max absolute difference: 3.57627869e-07
E       Max relative difference: 3.57627869e-07
E        x: array([[ 1.000000e+00+0.000000e+00j, -1.676381e-08+1.490116e-08j,
E                0.000000e+00+1.490116e-08j, -2.793968e-08-2.980232e-08j,
E               -2.235174e-08-1.862645e-08j, -2.095476e-08+5.587935e-09j,...
E        y: array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
E              [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
E              [0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],...

a          = array([[-0.11519074-0.22085421j,  0.4031076 +0.25545135j,
         0.20641464-0.02725353j,  0.04916074+0.12054591j,
  ....17945503j,  0.10257003-0.11285198j,
        -0.05898733+0.03267026j,  0.5694973 -0.06679238j]],
      dtype=complex64)
aTa        = array([[ 1.0000001e+00+0.0000000e+00j, -1.6763806e-08+1.4901161e-08j,
         0.0000000e+00+1.4901161e-08j, -2.793967...3484e-08-4.4703484e-08j,
        -8.3819032e-09-2.9802322e-08j,  9.9999994e-01+0.0000000e+00j]],
      dtype=complex64)
assert_sqr = True
atol       = 2.384185791015625e-07
rtol       = 0.0001
```

These tests are minor enough in failure that i'm uploading the artifact built to Anaconda.org

### LINKED ISSUE #136: Error Installing SciPy 1.5.2 on 64-Bit Windows 10 and Python 3.8
Issue:
Error Installing SciPy 1.5.2 on 64-Bit Windows 10 v. 1909 and Python 3.8.6
<details>

```
(new2) C:\Windows\system32>conda install -c conda-forge scipy=1.5.2
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: /
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - scipy=1.5.2 -> python[version='>=3.6,<3.7.0a0|>=3.7,<3.8.0a0']

Your python: python=3.8.6

If python is on the left-most side of the chain, that's the version you've asked for.
When python appears to the right, that indicates that the thing on the left is somehow
not available for the python version you are constrained to. Note that conda will not
change your python version to a different minor version unless you explicitly specify
that.

The following specifications were found to be incompatible with your system:

  - feature:/win-64::__cuda==11.1=0
  - feature:|@/win-64::__cuda==11.1=0

Your installed version is: 11.1
```
</details>

<br/>
Environment (<code>conda list</code>):
<details>

```
(new2) C:\Windows\system32>conda list
# packages in environment at C:\ProgramData\Anaconda3_8\envs\new2:
#
# Name                    Version                   Build  Channel
blas                      1.0                         mkl    conda-forge
ca-certificates           2020.6.20            hecda079_0    conda-forge
intel-openmp              2020.2                      254
libblas                   3.8.0           8_h8933c1f_netlib    conda-forge
libcblas                  3.8.0           8_h8933c1f_netlib    conda-forge
liblapack                 3.8.0           8_h8933c1f_netlib    conda-forge
m2w64-gcc-libgfortran     5.3.0                         6
m2w64-gcc-libs            5.3.0                         7
m2w64-gcc-libs-core       5.3.0                         7
m2w64-gmp                 6.1.0                         2
m2w64-libwinpthread-git   5.0.0.4634.697f757               2
mkl                       2020.2                      256
msys2-conda-epoch         20160418                      1
numpy                     1.19.1           py38h72c728b_0    conda-forge
openssl                   1.1.1h               he774522_0    conda-forge
python                    3.8.6           h60c2a47_0_cpython    conda-forge
python_abi                3.8                      1_cp38    conda-forge
sqlite                    3.33.0               he774522_1    conda-forge
tzdata                    2020b                h516909a_0    conda-forge
vc                        14.1                 h869be7e_1    conda-forge
vs2015_runtime            14.16.27012          h30e32a0_2    conda-forge

```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
(new2) C:\Windows\system32>conda info

     active environment : new2
    active env location : C:\ProgramData\Anaconda3_8\envs\new2
            shell level : 7
       user config file : C:\Users\name\.condarc
 populated config files : C:\Users\name\.condarc
          conda version : 4.8.5
    conda-build version : 3.20.3
         python version : 3.8.6.final.0
       virtual packages : __cuda=11.1
       base environment : C:\ProgramData\Anaconda3_8  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/win-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/win-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/win-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/msys2/win-64
                          https://repo.anaconda.com/pkgs/msys2/noarch
          package cache : C:\ProgramData\Anaconda3_8\pkgs
                          C:\Users\name\.conda\pkgs
                          C:\Users\name\AppData\Local\conda\conda\pkgs
       envs directories : C:\ProgramData\Anaconda3_8\envs
                          C:\Users\name\.conda\envs
                          C:\Users\name\AppData\Local\conda\conda\envs
               platform : win-64
             user-agent : conda/4.8.5 requests/2.24.0 CPython/3.8.6 Windows/10 Windows/10.0.18362
          administrator : True
             netrc file : None
           offline mode : False

```
</details>

### LINKED ISSUE #139: Python 3.9 builds for windows
Merging #138 only built for linux/osx but continues the 3.9-migration. However, all packages depending on scipy and building for windows will necessarily fail without a windows build (from either CF or anaconda). Last time, this was done manually, I believe (cf. #119).

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Fixes #80
Fixes #119
Fixes #136
Fixes #139

### CODE REVIEW

File: recipe/bld.bat [hunk 1]
```diff
@@ -1,17 +1,35 @@
-set COMP_DIR=C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows
+REM these re done automatically for openblas by numpy.distutils, but
```
Comment (h-vetinari) [inline recipe/bld.bat:1]:
```suggestion
REM these are done automatically for openblas by numpy.distutils, but
```

File: recipe/bld.bat [hunk 2]
```diff
@@ -1,17 +1,35 @@
...
-    set VS=vs2015
-) else (
-    set VS=vs2010
-)
+REM Add a file to load the fortran wrapper libraries in scipy/.libs
+rm scipy/_distributor_init.py
+cp %RECIPE_DIR%/_distributor_init.py scipy/_distributor_init.py
 
-call "%COMP_DIR%\bin\ifortvars.bat" %INTEL_ARCH% %VS%
+%PYTHON% -m pip install . -vv
```
Comment (h-vetinari) [inline recipe/bld.bat:27]:
```suggestion
%PYTHON% -m pip install . -vv
if %ERRORLEVEL% neq 0 exit 1
```

**Review by h-vetinari — COMMENTED** [review]
Very cool that you got this to build 🚀

File: recipe/meta.yaml [hunk 3]
```diff
@@ -20,7 +20,10 @@ requirements:
     - cython                                 # [build_platform != target_platform]
     - numpy                                  # [build_platform != target_platform]
     - pybind11                               # [build_platform != target_platform]
-    - {{ compiler('fortran') }}
+    - {{ compiler('fortran') }}   # [unix]
+    # WARNING: It's not recommended to use these MinGW compilers with python extensions
+    # numpy.distutils has a complex mechanism to facilitate mixing gfortran and MSVC
```
Comment (rgommers) [inline recipe/meta.yaml:25]: Maybe useful as a reference: https://pav.iki.fi/blog/2017-10-08/pywingfortran.html#building-python-wheels-with-fortran-for-windows

**Comment by rgommers** [comment]
Awesome, thanks so much @isuruf! Looks like this is all good to go for Python 3.6-3.8, passing the full test suite.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -42,9 +45,17 @@ test:
...
+    - curl      # [win]
+    - m2-unzip  # [win]
 {% if ver_major_minor == "1.5" %}
     - blas * netlib  # [ppc64le]
 {% endif %}
+  commands:
+    - curl -L -O https://github.com/lucasg/Dependencies/releases/download/v1.10/Dependencies_x64_Release.zip  # [win]
+    - unzip Dependencies_x64_Release.zip  # [win]
+    - Dependencies.exe -imports %SP_DIR%\scipy\linalg\_fblas.cp39-win_amd64.pyd  # [win]
+    - set "PATH=%SP_DIR%\scipy\.libs;%PATH%"  # [win]
```
Comment (isuruf) [inline recipe/meta.yaml:57]: I don't know why this is needed in python 3.9 windows. I even tried adding `os.add_dll_directory` in `_distributor_init.py`.
cc @rgommers, @mingwandroid

**Comment by Anthchirp** [comment]
win64 build job raising warnings with different libraries
```
WARNING (scipy,Lib/site-packages/scipy/linalg/cython_lapack.cp38-win_amd64.pyd): $RPATH/lib_blas_su.LV6ZZ35WNMXSLYOWI2AGMLM5ZSYBYIOZ.gfortran-win_amd64.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
(..)
WARNING (scipy,Lib/site-packages/scipy/optimize/_minpack.cp38-win_amd64.pyd): $RPATH/libchkder.YSCWF634KIJASKK5YRIPBQTJLCNZ66JU.gfortran-win_amd64.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
(..)
WARNING (scipy,Lib/site-packages/scipy/interpolate/_fitpack.cp38-win_amd64.pyd): $RPATH/libbispeu.RTWFDG7WJKRVJBKUFFCHOE7NCDHIJOUV.gfortran-win_amd64.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.
.. is this binary repackaging?
(..)
```
and ultimately failing in the test stage 
```python
===== testing package: scipy-1.5.2-py38h5f893b4_2 =====
running run_test.py
Traceback (most recent call last):
  File "C:\bld\scipy_1604186109765\test_tmp\run_test.py", line 13, in <module>
    import scipy
  File "C:\bld\scipy_1604186109765\_test_env\lib\site-packages\scipy\__init__.py", line 136, in <module>
    from . import _distributor_init
  File "C:\bld\scipy_1604186109765\_test_env\lib\site-packages\scipy\_distributor_init.py", line 61, in <module>
    WinDLL(os.path.abspath(filename))
  File "C:\bld\scipy_1604186109765\_test_env\lib\ctypes\__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
FileNotFoundError: Could not find module 'C:\bld\scipy_1604186109765\_test_env\lib\site-packages\scipy\.libs\libbanded5x.IEGDX6Z7J7WMHMZPXHFVNG23WKASH4R3.gfortran-win_amd64.dll' (or one of its dependencies). Try using the full path with constructor syntax.
```

**Comment by Anthchirp** [comment]
the files themselves are in `/.libs/` and pass validation, but it looks like they are not being found in there by the `.pyd`s?

File: recipe/meta.yaml [hunk 5]
```diff
@@ -42,6 +46,8 @@ test:
     - pytest
     - pytest-xdist
     - mpmath
+    - curl      # [win]
+    - m2-unzip  # [win]
```
Comment (h-vetinari) [inline recipe/meta.yaml:50]: These seem to be a left-over from debugging...

**Review by h-vetinari — COMMENTED** [review]
Awesome work, thanks so much! Found one more nit for fixing next time.

**Comment by tboerstad** [comment]
Thank you very much for your work, @isuruf!

**Comment by DanCodigaMWRA** [comment]
I use conda on Windows and like to avoid MKL (preferring to instead use openblas) because of the unwieldy size that MKL can give to executables created by, say, pyinstaller or nuitka.
But so far I have been unable to find any way to use scipy with anything non-MKL, via conda, on Windows. 
I had thought that issue #80 was intended to address that need (this could be me not understanding), and apparently this #146 fixes #80. 

** Am I understanding correctly, that as a result of #146 I now can get scipy without MKL on Windows using conda (I'm using "blas=*=openblas" in the yml file when creating a virtual env)?

**Comment by isuruf** [comment]
Have you tried using openblas? Can you please open a separate issue with info on what you tried and how it failed?

**Comment by DanCodigaMWRA** [comment]
I have tried with openblas, as I mentioned, but it was months ago. 
For now I just wanted to understand whether what I'm seeing here in #146 means that I should go try again. It sounds like I should. 
(No failure to report!)

**Comment by jakirkham** [comment]
If this was last tried months ago, would suggest retrying as this was merged Oct 31st. If you still run into issues, please file a new issue. Thanks! 😄

**Comment by DanCodigaMWRA** [comment]
Okay so I tried it and I believe it succeeded. This should be quite helpful-- can now avoid MKL while using scipy via conda on Windows. If I find any challenges or shortcomings with it later I will post a new issue, but for now no such need. Thank you for this important improvement!

File: recipe/meta.yaml (hunk 4 above)
Comment (rgommers) [reply recipe/meta.yaml]: Was just looking at this PR again, and I believe it's because of this conda bug: https://github.com/conda/conda/issues/10897

**Pull request merged**